### PR TITLE
Potential New Feature: Hierarchical Softmax

### PR DESCRIPTION
**Pull request by jmhessel**
(pull request version of issue #3143 , hoping to get some feedback as a pull request, though I doubt the code I currently have written is adequate. I don't want to go through doing PEP-8 things/squashing everything down into a single commit just yet, because there's still work to do.)
# What is hierarchical softmax used for?

Hierarchical softmax is an approximate version of softmax that is much faster to compute when the number of potential classes is high. In general, hierarchical softmax is used in language modeling tasks, where vocabulary sizes exceed ~10K. For very large vocabulary scenarios (think 1M word vocab) the normal softmax is computationally expensive. Two popular methods for approximating the softmax in cases of large vocabularies are negative sampling (a simplification of noise-constrastive estimation) and hierarchical softmax. The version I have implemented is the two-layer version, originally given [here](http://arxiv.org/abs/cs/0108006). This version is thought to be efficient for GPUs, in particular, because it decomposes the softmax into two reasonably sized matrix multiplications.
# How does hierarchical softmax work?

Hierarchical softmax can be thought of as a tree -- each leaf node of the tree is a class of softmax. Instead of computing a probability distribution over all possible classes given an input vector, hierarchical softmax computes the probability of a _single given class_ (generally the correct class) given an input vector. Each node on the path from the root to the correct leaf node is tasked with making a decision about which of its children it should go to. Each node learns a linear classifier to pick a child node given an input, and the total probability of the correct class is modeled as the product of the probabilities of the correct decision at each node from the root to the correct leaf node.

For a much better explanation, check out Hugo Larochelle's [video](https://www.youtube.com/watch?v=B95LTf2rVWM)
# What's the status of the implementation?

I have forked keras, and believe that I have a somewhat functional version of [two-layer hierarchical softmax](https://github.com/jmhessel/keras/blob/master/keras/layers/advanced_activations.py#L268-L395). However, it's pretty messy (see below).

I have tested it on a simple MNIST MLP classification model, and in a "large number of classes" case. In some cases, it runs 10x faster than normal softmax. In other cases, it's comparable or even slower than normal softmax. I wrote a [dummy task](https://github.com/jmhessel/keras/blob/master/examples/hsm_benchmark.py) script, which also demonstrates how to use the layer..
# Remaining questions
- Is this a feature keras wants to have in the main repo? Should I keep working on this?
- I added an additional objective function: "mean identity error," which basically ignores the targets, and simply takes a mean over the input tensor. This causes Theano to raise a warning about unused portions of the computation graph. Mean identity error is good for hierarchical softmax, because the layer outputs the negative log probability of the correct class -- this is something that should probably be minimized. The "right" solution might be to have models that require no labels (e.g. it could become legal to pass "None" as the target when using mean identity error; this could be useful for a variety of models outside of hierarchical softmax).
- Is my implementation too messy? It was a bit of a pain to write, and seems a bit inefficient, and I had to do some things in a relatively round-about way (e.g. the indexing operations are strange and messy). The messiness potentially derives from both my own coding skills, and from the limited set of functions supported by the backend (see below).
- Theano has an [implementation](https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/nnet.py#L2234-L2347) of hierarchical softmax, which would be ideal to use. Theano's HSM is made efficient by a specialized call to "sparse_block_dot," which is not in the keras backend because, to my knowledge, this operation doesn't exist in tensorflow. Many of the messy operations I had to write would be eliminated if there were access to this operation through keras. Would it be possible to support an operation only for one backend (this would allow for the use of theano's optimized implementation)? Or is that not worth it?

### CODE REVIEW

**Comment by nouiz** [comment]
I think you should use the Theano HSM for Theano, otherwise Theano will be
much slower then it should. If Tensorflow don't have an sparse_block_dot
operation, then in the Keras back-end, make an HSM function for Theano and
for Tensorflow. This way, you will have an interface that allow both
back-end to be supported. You could use your code in the tensorflow
back-end.

note, I didn't read the code of this PR.

**Comment by jmhessel** [comment]
I agree -- it seems silly to not use the optimized Theano implementation. Maybe I could add a warning that prints if you try to use the tensorflow backend with HSM?

To emphasize -- I do think this implementation is too messy to be merged as is; I mostly am just looking for feedback about the best direction to go from here.

File: keras/objectives.py [hunk 1]
```diff
@@ -11,6 +11,10 @@ def mean_absolute_error(y_true, y_pred):
     return K.mean(K.abs(y_pred - y_true), axis=-1)
 
 
+def mean_identity_error(y_true, y_pred):
+    return K.mean(y_pred, axis=-1)
```
Comment (bkj) [inline keras/objectives.py]: It's a hack, but changing to something like 

```
return K.mean(y_pred, axis=-1) + (0 * y_true)
```

may suppress the Theano warning that you're currently throwing.

**Comment by farizrahman4u** [comment]
I am doing something similar, whats the time difference you are observing?

File: keras/objectives.py (hunk 1 above)
Comment (jmhessel) [inline keras/objectives.py]: Thanks for the insight! I will consider adding in a trivial usage of the y_true labels to suppress the Theano warnings.

**Comment by jmhessel** [comment]
@farizrahman4u It heavily depends on the number of classes, etc. Given this naive implementation, I have seen speedups of up to 10x over normal softmax, or slowdowns of over 10x. Having the sparse block dot operation really seems to be key, but perhaps you've found a way around it?

**Comment by fchollet** [comment]
I think this is promising and it would be neat to have it in Keras. Ideally though, we would find a way to integrate it more cleanly with the Keras API, i.e. make it run with `fit(inputs, labels)`.

I think the following could be possible:

- `HierachicalSoftmax`, as a layer, returns `[input] + weights` in train mode, and returns probabilities in test mode.
- a `hierarchical_softmax` loss function takes as input the labels and the output of the `HS` layer (`[input] + weights`), and computes a loss value.

What do you think? The above is not great either, but it's still better that the current setup (and yes, I remember that I was the one who suggested the current setup in the first place), since it allows `fit(inputs, labels)`.

**Comment by farizrahman4u** [comment]
I already have a working hierarchical softmax layer which uses huffman trees. The user could provide a frequency table (e.g, if the task is predicting the probability of a word, common words like 'is', 'are', etc would have higher frequency) and the internal huffman tree would be built in such a way that frequent classes have shorter paths.. the problem is that I am getting a weird c++ error when the number of classes are too big(~100K)..Will fix it by this weekend.

**Comment by jmhessel** [comment]
Oh, that's awesome -- huffman trees are how word2vec et al. do it, so that's very cool. I decided to do the two-layer version because it seemed easier for the GPU-friendly libraries like Theano to work with that. How are you writing your HSM in the keras backend? Did you end up having to add TF/Theano ops? I have a few ideas about methods superior to the one in this PR, so I'm just curious at a high level.

I still think the depth-2 HSM tree is worth looking at, especially given that Theano has a nice implementation using sparse_block_dot.

**Comment by farizrahman4u** [comment]
The keras backend is pretty thorough now.. I use only K functions, no backend specific code. If there is anything I need, I simply PR them. I did a bit a bit of math and reduced the whole huffman tree traversal to a single dot product (and an addition).

**Comment by jmhessel** [comment]
@fchollet I can make those API changes, sure. I do want to make sure that the backend stuff is efficient enough first, though, if that's okay.

@farizrahman4u Gotcha. I don't see offhand how you were able to traverse the tree using a single dot product -- any chance your code is available to look at? What sorts of speedups are you observing?

**Comment by ftrier** [comment]
Hi,
how would I use the layer for prediction. I would always require two inputs and do not receive a probability distribution.

**Comment by dead** [comment]
Any updates on this? 

@fchollet 
I was trying to do what you said, but how can I get the output of the HS layer in the loss function?

**Comment by jmhessel** [comment]
@ftrier HSM often isn't used for prediction -- for many of the models trained with it, the internals of the model are generally the outputs, rather than the model itself (e.g. word embeddings). However, if you wanted to use HSM for prediction, I could see two reasonable methods. The first would be a brute-force approach, where you compute the probability of each class in the tree, e.g. to get the full probability distribution, you take your input vector x and compute the probability for each class y_i in a big loop. The second would be a tree search -- if you weren't interested in the whole distribution and were rather interested in simply the most likely output, you could probably come up with a smart way to traverse the tree from the root to the most likely leaf without evaluating each time.

@dead No updates on my end right now; I've been pretty slammed at work. My next step will be to change the API to what @fchollet mentioned, and put in the fast theano HSM implementation into the backend. If @farizrahman4u gets back with a nicer implementation in the backend, then that could be used for the tensorflow side of things (or, depending on its speed) both Theano and Tensorflow.

**Comment by farizrahman4u** [comment]
I am getting pretty good (>10x on training) speed on theano backend. Haven't tried tensorflow yet. Am still working on making it faster.

**Comment by farizrahman4u** [comment]
Yes, this is using pure keras.

**Comment by fchollet** [comment]
Do you have any update on this? By the way, this is similar to the CTC loss example we checked in: https://github.com/fchollet/keras/blob/master/examples/image_ocr.py#L433

I'm considering adding a built-in way to support this type of loss.

**Comment by jmhessel** [comment]
@fchollet No update at the moment -- I haven't had a chance to sit down in the past few weeks :-/ However, this is definitely on my radar and I plan to put in some work in the next week or so. I'll check out ctc loss.

**Comment by jmhessel** [comment]
The c2c loss looks like the same as the mean identity loss that was part of this PR way back -- so, that's an option, potentially.

**Comment by braingineer** [comment]
in case you all didn't seen this paper, from some folks at Facebook [Strategies for Training Large Vocabulary Neural Language Models](http://arxiv.org/abs/1512.04906).

> Welin Chen, David Grangier, Michael Auli
(Submitted on 15 Dec 2015)
Training neural network language models over large vocabularies is still computationally very costly compared to count-based models such as Kneser-Ney. At the same time, neural language models are gaining popularity for many applications such as speech recognition and machine translation whose success depends on scalability. We present a systematic comparison of strategies to represent and train large vocabularies, including softmax, hierarchical softmax, target sampling, noise contrastive estimation and self normalization. We further extend self normalization to be a proper estimator of likelihood and introduce an efficient variant of softmax. We evaluate each method on three popular benchmarks, examining performance on rare words, the speed/accuracy trade-off and complementarity to Kneser-Ney.

**Comment by jmhessel** [comment]
Okay -- I updated the HSM layer to use the theano backend. It runs in some cases and is pretty fast, but 3 major problems remain:

1. Tensorflow doesn't have an equivalent specialized HSM op. Perhaps @farizrahman4u's implementation could be used instead?
2. The op requires a "batch size" parameter. I can't figure out a way to have this abstracted away, so I had to add a "batch size" parameter to the layer itself. This is in-optimal for multiple reasons, not the least of which is that if your training data size is not a multiple of your batch size, it will crash. I can envision a few different solutions to this, but it's a bit dirty.
3. I have not updated the interface to @fchollet's new idea. I also haven't heard anything new about making this sort of loss (e.g. where the loss is computed as part of the model) easier to use, so I'm sticking with the "mean identity error" function I've been using thus far.

**Comment by jmhessel** [comment]
For now, though -- I could translate the built-in [theano HSM[(https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/nnet.py#L2234-L2347) into tensorflow. Sparse block dot will be difficult to replicate efficiently, but such is life.

**Comment by jmhessel** [comment]
And thanks @braingineer; I am aware of that paper! In fact, it is one I cite in the comments for HSM :-)

**Comment by around1991** [comment]
For problem 2, you could dynamically find out the batch_size to pass to h_softmax by doing input.shape[0] (or similar) - afaict all the Theano backend functions should work just fine with a Theano variable. Don't know how this would affect TF though.

**Comment by jmhessel** [comment]
@around1991 Done! Thanks. I can fix problem 3 tomorrow (or very shortly) and maybe @farizrahman4u can comment on problem 1? How goes that pure backend implementation?

**Comment by jmhessel** [comment]
@fchollet I was re-reading your idea about having a hierarchical softmax loss function and I'm not quite sure I understand. The speed of HSM is derived from the fact that the layer outputs a single value given a vector input and a desired label, so, the output of the HSM layer is simply the probability of the given class (generally: the correct one) given the input. I can certainly have the layer have different behavior in test mode by outputting the full softmax, so a distribution can be given over everything if a label is not available. I'll add that tomorrow.

**Comment by jmhessel** [comment]
For issue 1, too, I think there is some hope. I will take a closer look at all of the tf ops and see what I can do.

**Comment by farizrahman4u** [comment]
@jmhessel you can find the label with max prob, instead of finding the prob of a given label using  a Huffman tree.

**Comment by around1991** [comment]
@jmhessel: there's a backend flag K.learning_phase() which returns either 0 or 1 depending on whether you're testing or evaluating. I think @fchollet is definitely right: the best case scenario is to be able to call model.fit(inputs, labels), rather than have labels as one of the inputs. In my use case (generating text from a language model), not having to pass labels as inputs would be ideal.

I think what @fchollet was getting at what is that in training mode, you can defer execution of the backend K.hierarchical_softmax function to the loss function (which knows the labels already), and in test mode you can execute K.hierarchical_softmax inside the HierarchicalSoftmax layer and output the full probabilities.

Hope this helps
Kris

**Comment by jmhessel** [comment]
@around1991 That makes sense! I'll take another gradient step on this PR shortly :-)

**Comment by around1991** [comment]
I had a look at the Theano code underlying the hierarchical softmax method: it seems the main speedups are due to having optimized sparse block dot implementations in Theano (see http://deeplearning.net/software/theano/library/tensor/nnet/blocksparse.html), and also from decomposing a single softmax into two smaller multiplications.

I'm not sure whether Tensorflow has similarly optimized sparse dot backend ops, but it should definitely be possible to obtain some speedup from decomposing a single large softmax into two smaller softmaxes. This should be relatively easy to do purely in the backend, but it does mean that HS on Tensorflow might be a lot slower.

Instead of just commenting from the sidelines (which is mainly what I've been doing so far), I can have a go at implementing this myself though :). I'm a bit busy today and tomorrow (it's my birthday), but I should have some time on Sunday to have a look.

Kris

**Comment by around1991** [comment]
Ach, had a look at K.learning_phase(). It's a symbolic tensor, whose value is fed in as one of the inputs to the compiled function when you execute either the train or the test function. So there's no way to condition programatic behaviour on it.

Unless there's another transparent place that records whether you're in train/test mode in Keras somewhere, the current solution seems to be the best.

**Comment by jmhessel** [comment]
Good point. Batch norm seems to deal with this by having an extra set of functions in the backend that are highly specialized, but that probably isn't ideal in this situation.

Regardless, I'll keep the current configuration for now and see if I can replicate Theano's hierarchical softmax in the keras backend.

**Comment by farizrahman4u** [comment]
www.github.com/farizrahman4u/huffmax

**Comment by sunlylorn** [comment]
@farizrahman4u  huffmax cannot use with Keras Wrapper TimeDistributed for input_shape matching issue.

**Comment by Dejust** [comment]
How can we use TimeDistributed with HierarchicalSoftmax? Is there any hacks?

**Comment by jmhessel** [comment]
As a quick update, for people looking at this in the future: I was able to make HSM work with the theano backend in keras. However, the speed gains were not high enough for the use-cases I considered to justify sinking more time into this. To @Dejust : HSM doesn't currently work, but if you use the theano hsm op, I think any layer/wrapper within the model would work.

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.

Note: I still think that a hierarchical softmax layer would be great to have. This should be made much simpler by the changes in Keras 2.

**Comment by jiangnanhugo** [comment]
@jmhessel https://github.com/jiangnanhugo/lmkit/blob/master/lmkit/layers I just write a class-based hierarchical softmax (C_softmax.py) and tree-based hierarchical softmax (H_softmax.py) in theano. Hope it might help you integrate with keras. I implement it without using the sparse_block_dot function, which is messy to understand.

**Comment by jmhessel** [comment]
@jiangnanhugo Thanks for the info! I am not currently working on this at all, but that's good to know. I think, though, @nouiz mentioned that the hierarchical softmax op in theano should be used out-of-the-box. However, your symbolic HSM could probably be translated into keras backend code, and used from there.

Overall, though -- I think it would be really nice for keras to have alternative softmaxes for high-dimensional output spaces. For those who are interested, I currently am using [tensorflow's sampled softmax loss](https://www.tensorflow.org/api_docs/python/tf/nn/sampled_softmax_loss) which I successfully wrapped in a keras layer.

I still don't see exactly how this sort of layer could be put into keras though, i.e., one that outputs loss-of-the-true-class at train time, and the full softmax at test time.
